packages/core/src/network/network.ts

```diff
@@ -0,0 +1,361 @@
+  input: string;
+}
+
+export class AgentNetwork extends MastraBase {
+  state: NetworkState;
+  agents: Agent[];
+  model: LanguageModelV1;
+  current_state?: CurrentNetworkState;
+  instructions?: string;
+  routingAgent: Agent;
+
+  constructor(config: NetworkConfig) {
+    super({ component: 'AGENT', name: 'NETWORK' });
+    this.state = config.initialState || new NetworkState({});
+    this.agents = config.agents;
+    this.model = config.routingModel;
+    this.instructions = config.instructions;
+
+    this.routingAgent = new Agent({
+      name: config.name,
+      instructions: this.getInstructions(),
+      model: this.model,
+    });
+  }
+
+  setCurrentState(state: CurrentNetworkState) {
+    this.current_state = state;
+  }
+
+  getState() {
+    return this.current_state;
+  }
+
+  getInstructions() {
+    const agentDescriptions = this.agents.map(agent => ({
+      name: agent.name,
+      description: agent.instructions,
+    }));
+
+    return `
+            You are a router in a network of specialized agents. Your job is to decide which agent should handle the next step or if the task is complete.
+
+            Available agents as tool calls:
+            ${agentDescriptions.map(a => `- ${a.name.replace(/[^a-zA-Z0-9_-]/g, '_')}`).join('\n')}
+
+            ${this.instructions}
+
+            Based on the above information, decide which agent should be called next or if the task is complete.
+            
+            WORKFLOW:
+            1. ALWAYS start by using the getNetworkState tool to check the current state of the network, including any previous agent results
+            2. Assess the current state and determine if you need to call an agent or if the task is complete
+            3. If you need to call an agent, execute the agent as a tool call with the appropriate message
+               - Include relevant information from previous agent results in your message to the next agent
+               - Make sure to pass along important context from the original query
+            4. IMMEDIATELY after receiving the agent's response, use getNetworkState again to see the updated network state
+               - This step is MANDATORY - you must check the state after EACH agent call
+            5. Assess if the task is now complete or if another agent needs to be called
+            6. If the task is complete, provide a final comprehensive answer based on all the information gathered from the agents
+            7. If another agent is needed, execute that agent as a tool call
+            
+            IMPORTANT: You MUST call getNetworkState before and after EVERY agent call. This ensures each agent has access to the results of previous agents.
+        `;
+  }
+
+  buildNetworkTools() {
+    // Define the tool type for better type safety
+    type AgentTool = ReturnType<typeof createTool>;
+
+    // Create tools from agents with state tracking
+    const agentAsTools = this.agents.reduce<Record<string, AgentTool>>(
```

Comment:
Thought for a later day: during the demo earlier the multi-step agent calling was a little slow and it just occured to me we could give the router agent a parallel agent tool so it could message multiple agents, they could generate at the same time, and it could get the result back. ex input args where each key is the name of an agent and the value is the message to send.
```
{ agentA: "do x", agentB: "do y" }
```
Res would match:
```
{ agentA: "As a large language model...", agentB: "lol" }
```

Reply:
nice thing about this would be that this could be a single tool then. The router could choose to call one or more agents with it